spec_integration/facter_resolve_spec.rb

```diff
@@ -3,6 +3,53 @@
 require_relative 'spec_helper'
 
 describe Facter do
+  context 'when recursively calling facter in an external/custom fact' do
+    ext_dir = Dir.mktmpdir
+    ext_fact_path = File.join(ext_dir, 'extfact')
+
+    if LegacyFacter::Util::Config.windows?
+      ext_fact_path << '.ps1'
+      ext_fact_path.tr!('/', '\\')
+      script = <<~SCRIPTEND
+        bundle exec facter --external-dir #{ext_dir} os
+      SCRIPTEND
+    else
+      script = <<~POSIXSCRIPT
+        #!/usr/bin/env ruby
+        require 'json'
+        `bundle exec facter --external-dir #{ext_dir} os`.strip
+      POSIXSCRIPT
+    end
+    File.write(ext_fact_path, script)
+
+    FileUtils.chmod('+x', ext_fact_path)
+    it 'detects facter invocations recursively and stops external and custom facts' do
+      facter_entrypoints = [
+        proc { `bundle exec facter --external-dir #{ext_dir} os` },
+        proc { Facter.resolve("--external-dir #{ext_dir}") },
+        proc { Facter.resolve("os.family --external-dir #{ext_dir}") },
+        proc {
+          Facter::OptionStore.external_dir = [ext_dir]
+          Facter.to_hash
+        }
+      ]
+      time = Benchmark.measure do
+        # This iterates over the supported entry points that would attempt to resolve
+        # external facts; if any one of them descends into recursion, it will exceed
+        # the DEFAULT_EXECUTION_TIMEOUT; otherwise, each of these calls should be only
+        # be a few seconds.
+        facter_entrypoints.each do |entrypoint|
+          entrypoint.call
+          Facter.reset
+          Facter.clear
+          Facter::OptionStore.reset
+          LegacyFacter.clear
```

Comment:
We recently found some more global state when running `spec` unit tests. I'm not sure if the same is needed here https://github.com/puppetlabs/facter/commit/b78365653f7248327e28e2dc9aba3b7d5f6efb07#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94? Or in the `spec_integration/spec_helper.rb`?

Reply:
I used some debug output in the test to verify that the external fact was indeed being called repeatedly throughout this loop of entrypoints; without these, subsequent calls to facter within this single test would just use cached values and not exercise the external fact collection.